### DISCUSSION TITLE
rendering into a single plane in another pile

### DISCUSSION THREAD

**Discussion by joseluis** [dankamongmen/notcurses#2398] [Q&A] [answered]
How would you recommend going about having different piles with multiple planes & visuals, and flattening/rendering/blitting/merging each pile into a single plane/visual in another pile, while maintaning the original piles?

**Comment by dankamongmen** [comment, marked as answer]
so you have Pile A, Pile B, Pile C

Pile A has Plane 1, Pile B has Plane 2 and Plane 3, Pile C has Plane 4

and you want to retain all this, plus have new Pile D with Plane 5, which is a combination of the other 4?

```
planeD1 = ncplane_dup(plane1);
planeD2 = ncplane_dup(plane2);
planeD3 = ncplane_dup(plane3);
planeD4 = ncplane_dup(plane4); // have duplicated all four planes
ncplane_reparent(planeD1, planeD1); // make new pile with plane1 at root
ncplane_reparent(planeD2, planeD1);
ncplane_reparent(planeD3, planeD1);
ncplane_reparent(planeD4, planeD1); // all dups are now in their own pile
ncplane_mergedown_simple(planeD2, planeD1);
ncplane_mergedown_simple(planeD3, planeD1);
ncplane_mergedown_simple(planeD4, planeD1);
```

that ought do it?

i'm unsure whether you can `mergedown` across piles. i don't think you should be able to. let me go check.

**Reply by joseluis** [reply]
Ehm, close. I was thinking more along the lines of having a new Pile D with 4 planes, one for each pile's mergedown. But being able to further combine those planes wouldn't hurt.

**Reply by dankamongmen** [reply]
so then this gets you what you needed, right?

**Reply by joseluis** [reply]
Also, would this work with pixels in any case?

**Reply by joseluis** [reply]
> so then this gets you what you needed, right?

So IIUC, if `mergedown` works across piles that would be enough for what I need?

**Reply by dankamongmen** [reply]
i do not believe `mergedown()` works for sprite planes, no

**Reply by joseluis** [reply]
So if SourcePileA has 40 planes, and I want to reduce them to just one into a different pile, I have to duplicate each single plane into the (same) new pile, merge that new pile down, and then reparent that plane to the final pile. Repeat that for each source pile...

**Reply by dankamongmen** [reply]
```c
  if(src->sprite || dst->sprite){                                                                                                   
    logerror("can't merge sprixel planes\n");                                                                                       
    return -1;                                                                                                                      
  }                      
```

**Reply by dankamongmen** [reply]
well you're saying you want to keep the copies of the old planes, right? so yeah, you need duplicate them, because merging 2 planes results in 1 plane; you lose it if you don't dup it...maybe i'm not understanding you. what would you expect to be able to do?

**Reply by dankamongmen** [reply]
being able to merge two sprixel planes is a reasonable thing...let me make a bug on that. i don't see how a sprixel plane and otherwise would be merged.

**Reply by dankamongmen** [reply]
#2399 added

**Reply by dankamongmen** [reply]
wait, i might be full of crap. it looks like `ncplane_mergedown()` *doesn't* destroy either source plane. so no need to dup. and furthermore it looks like it works across piles, which i need to think about a bit, but it works now.

**Reply by joseluis** [reply]
> well you're saying you want to keep the copies of the old planes, right? so yeah, you need duplicate them, because merging 2 planes results in 1 plane; you lose it if you don't dup it...maybe i'm not understanding you. what would you expect to be able to do?

Well, on one side confirmation there was no way of duplicating a whole pile, for example. And the mergedown across piles things.

On the other side:

> i do not believe `mergedown()` works for sprite planes, no

Confirmation on the best aproach for dealing with sprite planes. I guess I have to blit them first, and then move those planes to the pile I want them, separately from any mergedown process that only affects text planes... right?

EDIT: Ok I see the new issue. I think I may have enough to keep going.

**Reply by dankamongmen** [reply]
there's no functionality for combining sprite planes right now, though you ought be able to e.g. stack partially transparent ones
